### DISCUSSION TITLE
Add a way to test pages in isolation

### DISCUSSION THREAD

**Discussion by freiksenet** [gatsbyjs/gatsby#20481] [Ideas / Feature Requests]
## Summary

Currently one can't test pages with graphql queries. It would have been great if they would be testable in isolation.

### Basic example

There are two main ways to do it. Either we add a programmatic way to create "gatsby context", which is basically all that is needed to execute Gatsby GraphQL and then pass that to the page component so that it's rendered. Second way would be to add and/or document way to mock graphql calls in Gatsby pages.

**Comment by herecydev** [comment]
@freiksenet I'm keen to pick this up. 

I had a little play trying to invoke this stuff manually and I think it's possible. It looks like at a minimum the following lines need running: https://github.com/gatsbyjs/gatsby/blob/603a501b1e3c4ca8b2e6edbc2fc1a3f071be6977/packages/gatsby/src/bootstrap/index.ts#L14-L25

At that point we should have a schema. Do you have some further comments/design ideas about how you would achieve this?

**Comment by herecydev** [comment]
My ideal end result would be something like the following:

```jsx
import { createGraphqlContext } from "gatsby"
import MyGraphQlPage from "./pages/myGraphQlPage"

describe("graphql tests", () => {
  beforeAll(async () => {
    // mock api requests that sourceNodes will make
    stubData()

    // runs sourceNodes, schemaCustomizations, etc
    const runner = await createGraphqlContext();

    // ideally extract the following log in some nice helper
    const { data } = runner(myGraphQlPage.query)
    render(<MyGraphQlPage data={data} />)
  });

  it("works", () => {
    screen.getByText("somegraphqldata")
  });
});

```
